api/src/main/java/org/pmiops/workbench/cohortbuilder/SubjectCounter.java

```diff
@@ -0,0 +1,88 @@
+package org.pmiops.workbench.cohortbuilder;
+
+import com.google.cloud.bigquery.QueryParameterValue;
+import com.google.cloud.bigquery.QueryRequest;
+import org.pmiops.workbench.cohortbuilder.querybuilder.FactoryKey;
+import org.pmiops.workbench.cohortbuilder.querybuilder.QueryParameters;
+import org.pmiops.workbench.model.SearchGroup;
+import org.pmiops.workbench.model.SearchGroupItem;
+import org.pmiops.workbench.model.SearchRequest;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides counts of unique subjects
+ * defined by the provided {@link SearchRequest}.
+ */
+@Service
+public class SubjectCounter {
+
+    private static final String COUNT_SQL_TEMPLATE =
+            "select count(distinct person_id) as count\n" +
+                    "  from `${projectId}.${dataSetId}.person` person\n" +
+                    " where 1 = 1\n";
+
+    private static final String UNION_TEMPLATE =
+            " union distinct\n";
+
+    private static final String INCLUDE_SQL_TEMPLATE =
+            "and person.person_id in (${includeSql})\n";
+
+    private static final String EXCLUDE_SQL_TEMPLATE =
+            "and not exists\n" +
+                    "(select 'x' from\n" +
+                    "(${excludeSql})\n" +
+                    "x where x.person_id = person.person_id)\n";
+
+    /**
+     * Provides counts of unique subjects
+     * defined by the provided {@link SearchRequest}.
+     */
+    public QueryRequest buildSubjectCounterQuery(SearchRequest request) {
+        Map<String, QueryParameterValue> params = new HashMap<>();
+        List<String> queryParts = new ArrayList<>();
+        String finalSql = COUNT_SQL_TEMPLATE;
+
+        // build query for included search groups
+        for (SearchGroup includeGroup : request.getIncludes()) {
+            finalSql = finalSql + INCLUDE_SQL_TEMPLATE;
+            for (SearchGroupItem includeItem : includeGroup.getItems()) {
+                QueryRequest queryRequest = QueryBuilderFactory
+                        .getQueryBuilder(FactoryKey.getType(includeItem.getType()))
+                        .buildQueryRequest(new QueryParameters()
+                                .type(includeItem.getType())
+                                .parameters(includeItem.getSearchParameters()));
+                params.putAll(queryRequest.getNamedParameters());
```

Comment:
Is it guaranteed that any queries which use the same name for a parameter mean the same thing by it, and will get the same value?

Reply:
Yes, i think that is a true statement.